### PR TITLE
Refactor SonarAnalysisContext - Use IsUnchanged for all scenarios

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Part of #6532

`IsUnchanged` was not properly invoked for Symbol-based and Compilation-based rules. As we don't know a specific tree at the time of registration. Their `sourceTree` (tree where the issue is really reported) should be also checked against `IsUnchanged`.

Splitting the reporting base classes also revealed a bunch of rules that were by mistake invoking `ReportIssue` from Symbol-based and Compilation-based rules without checking for the GeneratedCode condition.

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs [hunk 1]
```diff
@@ -34,9 +34,11 @@ public abstract class SonarAnalysisContextBase
 
     public abstract bool TryGetValue<TValue>(SourceText text, SourceTextValueProvider<TValue> valueProvider, out TValue value);
 
-    public bool ShouldAnalyze(GeneratedCodeRecognizer generatedCodeRecognizer, SyntaxTree tree, Compilation compilation, AnalyzerOptions options) =>    // FIXME: This thing has confusing name
-        !IsUnchanged(tree, compilation, options)        // FIXME: This needs to go to
-        && (ShouldAnalyzeGenerated(compilation, options) || !tree.IsGenerated(generatedCodeRecognizer, compilation));
+    /// <param name="tree">Tree to decide on. Can be null for Symbol-based and Compilation-based scenarios. And we want to analyze those too.</param>
+    /// <param name="generatedCodeRecognizer">When set, generated trees are analyzed only when langauge-specific 'analyzeGeneratedCode' configuration property is also set.</param>
+    public bool ShouldAnalyzeTree(SyntaxTree tree, Compilation compilation, AnalyzerOptions options, GeneratedCodeRecognizer generatedCodeRecognizer) =>
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs:39]: The method name should be `ShouldAnalyze`. I find `Tree` redundant. We already provide the syntax tree as a parameter.

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs:39]: We use the `TryGetValue` where possible (see few lines above). For this, I don't see how we could use it as it depends on `SourceText` as a key and we don't have it here.

File: analyzers/its/expected/akka.net/Akka.Tests.Shared.Internals--netstandard2.0-S3241.json [hunk 2]
```diff
@@ -1,17 +0,0 @@
-{
-"issues":  [
-{
-"id":  "S3241",
-"message":  "Change return type to 'void'; not a single caller uses the returned value.",
-"location":  {
-"uri":  "sources\akka.net\src\core\Akka.Tests.Shared.Internals\AsyncContext\AsyncContext.TaskQueue.cs",
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/its/expected/akka.net/Akka.Tests.Shared.Internals--netstandard2.0-S3241.json:7]: I was investigating it when doing the change and I remember it was puzzling on 1st look. I concluded it's not perfect, but it's intended given how the generated code recognition works. I wish I have left a comment on the PR 🤦 

The technical reasoning was that the `AnalysisContext` is a partial class. And the method is used in https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/its/sources/akka.net/src/core/Akka.Tests.Shared.Internals/AsyncContext/AsyncContext.cs . And this `AnalysisContext.cs` contains `[DebuggerNonUserCode]` attribute making it a generated file. Changes I did started (canonically correctly) detecting that file as generated and that caused this issue to disappear.

The controversy is that generated code recognizer considers the full file generated. But that's how it works and it's not part of this refactoring to change that behavior.

I unfortunately don't remember the exact condition that made this change.

File: analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContext.cs [hunk 3]
```diff
@@ -71,49 +71,51 @@ internal SonarAnalysisContext(AnalysisContext analysisContext, IEnumerable<Diagn
...
-        analysisContext.RegisterCodeBlockStartAction<TSyntaxKind>(c => Execute<SonarCodeBlockStartAnalysisContext<TSyntaxKind>, CodeBlockStartAnalysisContext<TSyntaxKind>>(new(this, c), action));
+        analysisContext.RegisterCodeBlockStartAction<TSyntaxKind>(c => Execute<SonarCodeBlockStartAnalysisContext<TSyntaxKind>, CodeBlockStartAnalysisContext<TSyntaxKind>>(new(this, c), action, c.CodeBlock.SyntaxTree));
 
     public void RegisterCodeBlockStartActionInNonGenerated<TSyntaxKind>(GeneratedCodeRecognizer generatedCodeRecognizer, Action<SonarCodeBlockStartAnalysisContext<TSyntaxKind>> action)
         where TSyntaxKind : struct =>
-        analysisContext.RegisterCodeBlockStartAction<TSyntaxKind>(c => Execute<SonarCodeBlockStartAnalysisContext<TSyntaxKind>, CodeBlockStartAnalysisContext<TSyntaxKind>>(new(this, c), action, generatedCodeRecognizer));    // FIXME: Rename generatedCodeRecognizer everywhere
+        analysisContext.RegisterCodeBlockStartAction<TSyntaxKind>(c => Execute<SonarCodeBlockStartAnalysisContext<TSyntaxKind>, CodeBlockStartAnalysisContext<TSyntaxKind>>(new(this, c), action, c.CodeBlock.SyntaxTree, generatedCodeRecognizer));
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContext.cs:80]: This one is a bit tricky to spot. We could have provided it via context for `CodeBlock`, `CodeBlockStart`, `Node` and `Tree` variants, as the context has unique tree that we definitely know it's the correct one.

The problem is with `Compilation`, `CompilationStart` and `Symbol` contexts that do not know the correct tree at the time of registration. They only know the tree at the time of reporting. On the context level, they are using the `.FirstOrDefault()` shortcut that might not return the correct one (for `Symbol`) and will definitely return totally unrelated tree for `Compilation` and `CompilationStart`.

That's why the parameter is called `sourceTree` (not `syntaxTree`) and has extra documentation on it. As it was something very unusual.

File: analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContext.cs [hunk 4]
```diff
@@ -71,49 +71,51 @@ internal SonarAnalysisContext(AnalysisContext analysisContext, IEnumerable<Diagn
...
+    private void Execute<TSonarContext, TRoslynContext>(TSonarContext context, Action<TSonarContext> action, SyntaxTree sourceTree, GeneratedCodeRecognizer generatedCodeRecognizer = null)
         where TSonarContext : SonarAnalysisContextBase<TRoslynContext>
     {
         // For each action registered on context we need to do some pre-processing before actually calling the rule.
         // First, we need to ensure the rule does apply to the current scope (main vs test source).
         // Second, we call an external delegate (set by legacy SonarLint for VS) to ensure the rule should be run (usually
         // the decision is made on based on whether the project contains the analyzer as NuGet).
         if (context.HasMatchingScope(supportedDiagnostics)
-            && (generatedCodeRecognizer is null || context.ShouldAnalyze(generatedCodeRecognizer))
+            && context.ShouldAnalyzeTree(sourceTree, context.Compilation, context.Options, generatedCodeRecognizer)
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContext.cs:118]: We need `sourceTree`, see the previous comment and documentation of the parameter.

File: analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs:39]: The method was originally called `ShouldAnalyze` and it confused us during incremental PR analysis. I like the `Tree` suffix here to highlight that fact. While the tree is first argument, the method also receives `Compilation` and `AnalyzerOptions` and it's not clear that they are secondary noise.

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM! Rebase needed.

**Pull request merged**